### ISSUE TITLE
Moving to an empty UniqueHandle causes segfault.

### ISSUE THREAD

**Issue by CarpeNecopinum** [KhronosGroup/Vulkan-Hpp#163]
In my application I'm creating CommandPools and adding them to an unordered_map like this:
```cpp
mPools[index] = mDevice->createCommandPoolUnique({<flags...>, index});
```
During the move to the default-constructed (thus empty) `mPools[index]` the program segfaults. 

As a work-around I found
```cpp
void reset( Type const& value = Type() )
    {
      if ( m_value != value )
      {
        if ( m_value != Type() ) this->destroy( m_value );
        m_value = value;
      }
    }
```
(in the UniqueHandle class).

**Comment by asuessenbach** [comment]
Hi CarpeNecopinum,

could you please provide some more details on how to reproduce that issue? I tried but did not get any error.

Thanks,

   Andreas

**Comment by CarpeNecopinum** [comment]
I created a gist with a minimal-ish example that lets the error surface:
[https://gist.github.com/CarpeNecopinum/4e364076a648a7585a3161e8b874c926](https://gist.github.com/CarpeNecopinum/4e364076a648a7585a3161e8b874c926)
(pretty self-contained, can be compiled with just `g++ main.cc -lvulkan -lglfw -I. -g` on my system)

This is derived from one of the vulkan-tutorial.com examples (just so I have some code that produces a VkDevice). The code I added is
```cpp
QueueFamilyIndices indices = findQueueFamilies(physicalDevice);
auto dev = vk::Device(device);
std::unordered_map<int,vk::UniqueCommandPool> pools;
pools[indices.graphicsFamily] = dev.createCommandPoolUnique(vk::CommandPoolCreateInfo{
        vk::CommandPoolCreateFlagBits::eResetCommandBuffer,
        uint32_t(indices.graphicsFamily)});
```
inside the `initVulkan` method of the example.

**Comment by Kaosumaru** [comment]
Hi, I wanted to check out Vulkan-Hpp and also stumbled on this issue. I can give even smaller repro code.

    {
        vk::UniqueDebugReportCallbackEXT callback;
    }

I'm using `vulkan_ext.c` to load extensions. Since `m_value` in `UniqueHandle` is null, `this->destroy( m_value );` call in `~UniqueHandle` calls `pfn_vkDestroyDebugReportCallbackEXT(instance, nullptr, nullptr)` which causes segfault under windows.

I don't know if assumption is that vkDestroyDebugReportCallbackEXT should check if argument is empty, and `vulkan_ext.c` is inappropriate, or this is a silly bug in  `UniqueHandle`?

**Comment by mtavenrath** [comment]
That's a bug in UniqueHandle. Nearly all Vulkan handles allow vkDestroy(VK_NULL_HANDLE) and thus no check is required. 
vkDestroyDebugReportCallbackEXT deviates from the other destroy functions and does not allow a VK_NULL handle .This has to be fixed in Vulkan-Hpp.

**Comment by asuessenbach** [comment]
Fixed by #172.
Closing.
